File: go/uri/manifest.go
// SPDX-License-Identifier: MIT

package uri

import (
	"errors"
	"strings"
)

// HandlerDescriptor describes a single handler an app exposes on its
// aether:// URI surface.
//
// A handler is identified by its first path segment (the Name) plus an
// optional path template that captures route parameters. The router matches
// an incoming URI's HandlerName + path against a manifest of these and
// dispatches accordingly.
//
// Path template syntax
//
//	"content/{hash}"             // matches /content/abc
//	"watch/{sessionId}/join"     // matches /watch/123/join
//	"profile"                    // matches /profile exactly
//	"profile/avatar"             // matches /profile/avatar exactly
type HandlerDescriptor struct {
	// Name is the handler name — the first path segment (e.g. "content",
	// "stream").
	Name string

	// PathTemplate is the route template applied AFTER the handler name
	// (e.g. "{hash}", "{sessionId}/join"). Empty for a root handler that
	// matches exactly the handler name with no extra segments.
	PathTemplate string

	// ExpectedQueryKeys lists query keys the handler expects (informational
	// — not enforced by the manifest).
	ExpectedQueryKeys []string

	// Description is a human-readable description for diagnostics + docs.
	Description string
}

// Validate reports an error if the descriptor is missing a handler name.
func (d HandlerDescriptor) Validate() error {
	if strings.TrimSpace(d.Name) == "" {
		return errors.New("aether-uri: handler name is required")
	}
	return nil
}

// Match attempts to match the descriptor's template against an incoming URI
// path (the path component of a URI — already percent-decoded). Returns the
// captured route parameters and true on success, or (nil, false) on no match.
func (d HandlerDescriptor) Match(path string) (map[string]string, bool) {
	var templateSegs []string
	if d.PathTemplate == "" {
		templateSegs = []string{d.Name}
	} else {
		templateSegs = strings.Split(d.Name+"/"+strings.TrimLeft(d.PathTemplate, "/"), "/")
	}
	pathSegs := strings.Split(path, "/")
	if len(templateSegs) != len(pathSegs) {
		return nil, false
	}

	captures := map[string]string{}
	for i, t := range templateSegs {
		p := pathSegs[i]
		if len(t) >= 2 && t[0] == '{' && t[len(t)-1] == '}' {
			captures[t[1:len(t)-1]] = p
			continue
		}
		if t != p {
			return nil, false
		}
	}
	return captures, true
}

// ResolvedRoute is the result of resolving a URI against a HandlerManifest —
// the matched descriptor plus any captured route parameters.
type ResolvedRoute struct {
	// Handler is the descriptor that matched.
	Handler HandlerDescriptor

	// Captures holds the captured route parameters, keyed by template
	// placeholder name. Empty for handlers with no placeholders.
	Captures map[string]string
}

// HandlerManifest is an app's complete aether:// handler manifest — the set
// of routes the app accepts. Each app registers exactly one manifest at
// startup; the router dispatches against it.
type HandlerManifest struct {
	// AppID is the owning app's identifier (e.g. "aether.media",
	// "aether.txtme"). Reverse-DNS style.
	AppID string

	// Handlers is the ordered list of registered handler descriptors.
	// Order is significant for Resolve — the first matching descriptor wins.
	Handlers []HandlerDescriptor
}

// NewHandlerManifest constructs a HandlerManifest. Returns an error if the
// AppID is empty or any descriptor in handlers is invalid.
func NewHandlerManifest(appID string, handlers []HandlerDescriptor) (*HandlerManifest, error) {
	if strings.TrimSpace(appID) == "" {
		return nil, errors.New("aether-uri: appID is required")
	}
	for i, h := range handlers {
		if err := h.Validate(); err != nil {
			return nil, errFromIndex(i, err)
		}
	}
	// Copy to avoid aliasing.
	cp := make([]HandlerDescriptor, len(handlers))
	copy(cp, handlers)
	return &HandlerManifest{AppID: appID, Handlers: cp}, nil
}

// errFromIndex prefixes err with the offending handler index.
func errFromIndex(i int, err error) error {
	if err == nil {
		return nil
	}
	return errors.New("aether-uri: handler at index " + itoa(i) + ": " + err.Error())
}

// itoa is a tiny dependency-free int-to-string helper for small indices.
func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

// Resolve walks the manifest's handlers in order and returns the first one
// whose template matches uri's path, together with its captured parameters.
// Returns (nil, false) if no handler matches or if uri is invalid.
func (m *HandlerManifest) Resolve(u URI) (*ResolvedRoute, bool) {
	if m == nil || !u.IsValid() {
		return nil, false
	}
	handlerName := u.HandlerName()
	for _, h := range m.Handlers {
		if h.Name != handlerName {
			continue
		}
		captures, ok := h.Match(u.Path)
		if !ok {
			continue
		}
		return &ResolvedRoute{Handler: h, Captures: captures}, true
	}
	return nil, false
}

File: rust/src/uri/manifest.rs
// SPDX-License-Identifier: MIT

//! Handler manifest — the contract between an app and the URI router.
//!
//! Each app describes the `aether://` paths it accepts via a
//! [`HandlerManifest`] containing a list of [`HandlerDescriptor`]s. The
//! [`Router`](crate::uri::Router) matches an incoming URI against this
//! manifest and dispatches to the registered callback.
//!
//! # Path template syntax
//!
//! ```text
//! "content/{hash}"             // matches /content/abc
//! "watch/{sessionId}/join"     // matches /watch/123/join
//! "profile"                    // matches /profile exactly
//! "profile/avatar"             // matches /profile/avatar exactly
//! ""                           // matches /<handlerName> exactly
//! ```

use std::collections::BTreeMap;

use super::uri::AetherUri;

/// Describes a single handler an app exposes on its `aether://` URI surface.
#[derive(Clone, Debug)]
pub struct HandlerDescriptor {
    /// Handler name — the first path segment (e.g. `"content"`, `"stream"`).
    pub name: String,

    /// Path template (e.g. `"content/{hash}"`). Empty matches `/<name>` exactly.
    ///
    /// Note: the template does NOT include the handler-name prefix — that is
    /// implied by [`name`](Self::name). See the module-level docs for examples.
    pub path_template: String,

    /// Informational: query keys the handler expects. Not enforced by the
    /// router; surfaced for documentation and developer tooling.
    pub expected_query_keys: Vec<String>,

    /// Human-readable description for diagnostics and generated docs.
    pub description: String,
}

impl HandlerDescriptor {
    /// Convenience constructor.
    pub fn new(name: impl Into<String>, path_template: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            path_template: path_template.into(),
            expected_query_keys: Vec::new(),
            description: String::new(),
        }
    }

    /// Matches an incoming URI's path against this descriptor's template.
    ///
    /// Returns the captured route parameters on success, or `None` on no match.
    /// The descriptor's [`name`](Self::name) is prepended to the template
    /// before matching.
    pub fn match_path(&self, path: &str) -> Option<BTreeMap<String, String>> {
        // Build the full template (handler name + optional template suffix).
        let full_template: String = if self.path_template.is_empty() {
            self.name.clone()
        } else {
            let trimmed = self.path_template.trim_start_matches('/');
            format!("{}/{}", self.name, trimmed)
        };

        let template_segs: Vec<&str> = full_template.split('/').collect();
        let path_segs: Vec<&str> = path.split('/').collect();
        if template_segs.len() != path_segs.len() {
            return None;
        }

        let mut captures: BTreeMap<String, String> = BTreeMap::new();
        for (t, p) in template_segs.iter().zip(path_segs.iter()) {
            // {param} placeholders capture the path segment.
            if t.len() >= 2 && t.starts_with('{') && t.ends_with('}') {
                let key = &t[1..t.len() - 1];
                // Keys are case-insensitive — store lower-case to match the C# semantics.
                captures.insert(key.to_ascii_lowercase(), (*p).to_string());
            } else if t != p {
                return None;
            }
        }
        Some(captures)
    }
}

/// An app's complete `aether://` handler manifest.
///
/// Each app registers exactly one manifest at startup; the
/// [`Router`](crate::uri::Router) dispatches against it.
#[derive(Clone, Debug)]
pub struct HandlerManifest {
    /// The owning app's identifier (e.g. `"aether.media"`, `"aether.txtme"`).
    pub app_id: String,

    /// All registered handler descriptors. Order matters: the first match wins.
    pub handlers: Vec<HandlerDescriptor>,
}

impl HandlerManifest {
    /// Creates a new manifest.
    pub fn new(app_id: impl Into<String>, handlers: Vec<HandlerDescriptor>) -> Self {
        Self { app_id: app_id.into(), handlers }
    }

    /// Resolves an incoming URI against this manifest.
    ///
    /// Returns `(handler_index, captures)` if a descriptor matched, or `None`
    /// otherwise. The index is into the [`handlers`](Self::handlers) vector;
    /// returning an index rather than a reference keeps the API
    /// lifetime-free, which simplifies router state-management.
    pub fn resolve(&self, uri: &AetherUri) -> Option<(usize, BTreeMap<String, String>)> {
        for (i, h) in self.handlers.iter().enumerate() {
            if h.name != uri.handler_name() {
                continue;
            }
            if let Some(captures) = h.match_path(uri.path()) {
                return Some((i, captures));
            }
        }
        None
    }
}
